lms/djangoapps/grades/new/subsection_grade.py

```diff
@@ -135,29 +94,40 @@ def load_from_data(self, model, course_structure, scores_client, submissions_sco
             section=self.display_name,
             module_id=self.location,
         )
+        self._log_event(log.warning, u"init_from_model", user)
 
-    def __unicode__(self):
+    @classmethod
+    def bulk_create_models(cls, student, subsection_grades, course_key):
         """
-        Provides a unicode representation of the scoring
-        data for this subsection. Used for logging.
+        Saves the subsection grade in a persisted model.
         """
-        return u"SubsectionGrade|total: {0}/{1}|graded: {2}/{3}|location: {4}|display name: {5}".format(
-            self.all_total.earned,
-            self.all_total.possible,
-            self.graded_total.earned,
-            self.graded_total.possible,
-            self.location,
-            self.display_name
+        return PersistentSubsectionGrade.bulk_create_grades(
+            [subsection_grade._persisted_model_params(student) for subsection_grade in subsection_grades],  # pylint: disable=protected-access
+            course_key,
         )
 
+    def create_model(self, student):
+        """
+        Saves the subsection grade in a persisted model.
+        """
+        self._log_event(log.info, u"create_model", student)
+        return PersistentSubsectionGrade.create_grade(**self._persisted_model_params(student))
+
+    def update_or_create_model(self, student):
+        """
+        Saves or updates the subsection grade in a persisted model.
+        """
+        self._log_event(log.info, u"update_or_create_model", student)
```

Comment:
Better to use a unicode literal, though.  If you use a bytestring, then python has to convert it to a unicode string to merge it with the format specifier.  Then if your string has non-ascii characters in it, you'll get a UnicodeDecodeError.  If you use a unicode object, no conversion is required.

In general, you should think of `unicode` as representing text (a sequence of characters), and `str` as representing a sequence of bytes.  If you don't explicitly need bytes, you should default to unicode.  

The general recommended pattern is when you receive a bytestring, whether from reading data from a file or network socket or from an api that only provides bytes, turn it into unicode as soon as possible. Leave it unicode throughout your code until 1) you need to write it back to disk or the network or 2) you have to send it to a bytes-only api.  Django takes care of converting request and response data to bytes behind the scenes, so that takes care of a lot of the problem.  

The one exception is if you're dealing with binary data.  Then, of course, you leave it binary.